src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,197 @@
+                }
+
+                // This is the thread which will write to _instance, so this will act as an atomic read
+                var instance = _instance;
+
+                // Set the instance back to its default value, so in case someone forgets to dispose of one of the
+                // counted references the finalizer of the underlying disposable object will have a chance to clean up
+                // the unmanaged resources as soon as possible.
+                Interlocked.MemoryBarrier();
+                _instance = default(T);
+
+#if !DEBUG
+                var decrementedValue = Interlocked.Decrement(ref referenceCount.Value);
+                if (decrementedValue == 0)
+                {
+                    instance.Dispose();
+                }
+#else
+                while (true)
+                {
+                    var currentValue = Volatile.Read(ref referenceCount.Value);
+                    Debug.Assert(currentValue > 0, "Dispose should have protected itself against races.");
+
+                    if (Interlocked.CompareExchange(ref referenceCount.Value, currentValue - 1, currentValue) == currentValue)
+                    {
+                        if (currentValue == 1)
+                        {
+                            // Reference count hit 0 for this call
+                            instance.Dispose();
+                        }
+
+                        break;
+                    }
+                }
+#endif
+            }
+
+            /// <summary>
+            /// Represents a weak reference to a <see cref="ReferenceCountedDisposable{T}"/> which is capable of
+            /// obtaining a new counted reference up until the point when the object is no longer accessible.
+            /// </summary>
+            public struct WeakReference
+            {
+                /// <summary>
+                /// DO NOT DISPOSE OF THE TARGET.
+                /// </summary>
+                private readonly WeakReference<ReferenceCountedDisposable<T>> _instance;
+
+                public WeakReference(ReferenceCountedDisposable<T> reference)
+                    : this()
+                {
+                    var instance = reference._instance;
+                    Interlocked.MemoryBarrier();
+                    var referenceCount = Volatile.Read(ref reference._referenceCount);
+                    if (referenceCount == null)
+                    {
+                        // The specified reference is already not valid.
+                        return;
+                    }
+
+                    if (referenceCount.Value == 0)
+                    {
+                        // We were able to read the reference, but it's already disposed.
+                        return;
+                    }
+
+                    var innerReference = new ReferenceCountedDisposable<T>(instance, referenceCount);
+                    _instance = new WeakReference<ReferenceCountedDisposable<T>>(innerReference);
```

Comment:
i don't really understand how this can work.  You just made a new 'innerReference', and you are now holding onto it with a .Net WeakReference.  it doesn't look like anything is necessairly pointing at innerReference through a strong reference.  So i don't see how the first GC doesn't just clean this up immediately.